Sources/_StringProcessing/Algorithms/Matching/Matches.swift

```diff
@@ -183,13 +183,155 @@ extension BidirectionalCollection {
 
 // MARK: Regex algorithms
 
+@available(SwiftStdlib 5.7, *)
+struct RegexMatchesCollection<Output> {
+  let input: Substring
+  let regex: Regex<Output>
+  let startIndex: Index
+  
+  init(base: Substring, regex: Regex<Output>) {
+    self.input = base
+    self.regex = regex
+    self.startIndex = base.firstMatch(of: regex).map(Index.match) ?? .end
+  }
+}
+
+@available(SwiftStdlib 5.7, *)
+extension RegexMatchesCollection: Sequence {
+  /// Returns the index to start searching for the next match after `match`.
+  fileprivate func searchIndex(after match: Regex<Output>.Match) -> String.Index? {
+    if !match.range.isEmpty {
+      return match.range.upperBound
+    }
+    
+    // If the last match was an empty match, advance by one position and
+    // run again, unless at the end of `input`.
+    if match.range.lowerBound == input.endIndex {
+      return nil
+    }
+    
+    switch regex.initialOptions.semanticLevel {
+    case .graphemeCluster:
+      return input.index(after: match.range.upperBound)
+    case .unicodeScalar:
+      return input.unicodeScalars.index(after: match.range.upperBound)
+    }
+  }
+
+  struct Iterator: IteratorProtocol {
+    let base: RegexMatchesCollection
+    
+    // Because `RegexMatchesCollection` eagerly computes the first match for
+    // its `startIndex`, the iterator can use that match for its initial
+    // iteration. For subsequent calls to `next()`, this value is `false`, and
+    // `nextStart` is used to search for the next match.
+    var initialIteration = true
+    var nextStart: String.Index?
+    
+    init(_ matches: RegexMatchesCollection) {
+      self.base = matches
+      self.nextStart = base.startIndex.match.flatMap(base.searchIndex(after:))
+    }
+    
+    mutating func next() -> Regex<Output>.Match? {
+      // Initial case with pre-computed first match
+      if initialIteration {
+        initialIteration = false
+        return base.startIndex.match
+      }
+      
+      // `nextStart` is `nil` when iteration has completed
+      guard let start = nextStart else {
+        return nil
+      }
+      
+      // Otherwise, find the next match (if any) and compute `nextStart`
+      let match = try! base.regex.firstMatch(in: base.input[start...])
```

Comment:
Should this be `try!` or `try?`?

Reply:
`try?` seems like it would be fine, here.